File: java/src/main/java/be/steinsomers/bron_kerbosch/UndirectedGraph.java
package be.steinsomers.bron_kerbosch;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public record UndirectedGraph(List<Set<Integer>> adjacencies) {
    public UndirectedGraph {
        for (int v = 0; v < adjacencies.size(); ++v) {
            for (int w : adjacencies.get(v)) {
                assert v != w;
                assert adjacencies.get(w).contains(v);
            }
        }
    }

    public int order() {
        return adjacencies.size();
    }

    public int size() {
        int total = adjacencies.stream().mapToInt(Set::size).sum();
        assert total % 2 == 0;
        return total / 2;
    }

    public int degree(int node) {
        return adjacencies.get(node).size();
    }

    public Set<Integer> neighbours(int node) {
        return Collections.unmodifiableSet(adjacencies.get(node));
    }

    public Stream<Integer> connectedVertices() {
        return IntStream.range(0, order()).filter(v -> degree(v) > 0).boxed();
    }

    public int maxDegreeVertex() {
        return IntStream.range(0, order()).boxed()
                .max(Comparator.comparingInt(this::degree))
                .orElseThrow();
    }
}

File: kotlin/src/main/be/steinsomers/bron_kerbosch/UndirectedGraph.kt
package be.steinsomers.bron_kerbosch

fun isLoopFree(adjacencies: List<Set<Int>>): Boolean {
    return adjacencies.asSequence().filterIndexed { i, neighbours -> neighbours.contains(i) }
        .none()
}

fun areNeighboursReciprocalIn(adjacencies: List<Set<Int>>, v: Int, neighbours: Set<Int>): Boolean {
    return neighbours.all { w -> adjacencies[w].contains(v) }
}

fun isSymmetric(adjacencies: List<Set<Int>>): Boolean {
    return adjacencies.foldIndexed(true) { index, valid, neighbours ->
        valid && areNeighboursReciprocalIn(adjacencies, index, neighbours)
    }
}

data class UndirectedGraph(private val adjacencies: List<Set<Int>>) {
    val size: Int
    val maxDegree: Int

    init {
        Debug.assert { isLoopFree(adjacencies) }
        Debug.assert { isSymmetric(adjacencies) }
        maxDegree = if (adjacencies.any()) adjacencies.maxOf(Set<Int>::size) else 0
        val totalDegree = adjacencies.sumOf(Set<Int>::size)
        assert(totalDegree % 2 == 0)
        size = totalDegree / 2
    }

    val order: Int
        get() {
            return adjacencies.size
        }

    fun degree(vertex: Int): Int {
        return adjacencies[vertex].size
    }

    fun hasDegree(vertex: Int): Boolean {
        return adjacencies[vertex].isNotEmpty()
    }

    fun neighbours(vertex: Int): Set<Int> {
        return adjacencies[vertex]
    }

    fun connectedVertices(): Sequence<Int> {
        return (0..<order).asSequence().filter(this::hasDegree)
    }

    fun maxDegreeVertices(): Sequence<Int> {
        return (0..<order).asSequence().filter { v -> degree(v) == maxDegree }
    }
}
